cpp/src/io/csv/csv_writer.cu

```diff
@@ -0,0 +1,284 @@
+    const char* terminator = "\n";
+    if( args->line_terminator )
+        terminator = args->line_terminator;
+    const char* true_value = (args->true_value ? args->true_value : "true");
+    const char* false_value = (args->false_value ? args->false_value : "false");
+
+    // check for issues here
+    if( filepath==0 )
+    {
+        std::cerr << "write_csv: filepath not specified\n";
+        return GDF_INVALID_API_CALL;
+    }
+    if( count==0 )
+    {
+        std::cerr << "write_csv: num_cols is required\n";
+        return GDF_INVALID_API_CALL;
+    }
+    if( columns==0 )
+    {
+        std::cerr << "write_csv: invalid data values\n";
+        return GDF_INVALID_API_CALL;
+    }
+
+    // check all columns are the same size
+    for( int idx=0; idx < args->num_cols; ++idx )
+    {
+        gdf_column* col = args->data[idx];
+        NVStrings* strs = nullptr;
+        //printf("%d: %p %s %d %d\n", idx, col, col->col_name, (int)col->dtype, (int)col->size);
+        switch( col->dtype )
+        {
+            case GDF_INT8:
+            case GDF_INT32:
+            case GDF_INT64:
+            case GDF_FLOAT32:
+            case GDF_FLOAT64:
+            case GDF_DATE64:
+                if( (unsigned int)col->size != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << (int)col->size << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            case GDF_STRING:
+                strs = (NVStrings*)col->data;
+                if( strs->size() != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << strs->size() << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            default:
+                std::cerr << "write_csv: unknown column type: " << (int)col->dtype << "\n";
+                return GDF_UNSUPPORTED_DTYPE;
+        }
+    }
+
+    // check the file can be written
+    FILE* fh = fopen(filepath,"wb");
+    if( !fh )
+    {
+        std::cerr << "write_csv: file [" << filepath << "] could not be opened\n";
+        return GDF_FILE_ERROR;
+    }
+
+    //
+    // It would be good if we could chunk this.
+    // Use the rows*count calculation and a memory threshold to
+    // output a subset of rows at a time instead of the whole thing at once.
+    // The entire CSV must fit in CPU memory before writing it out.
+    //
+    // compute sizes we need
+    // build a matrix of pointers
+    std::unique_ptr<int> pdatalens(new int[rows*count]); // matrix
+    int* datalens = pdatalens.get();
+    size_t memsize = 0;
+    for( unsigned int idx=0; idx < count; ++idx )
+    {
+        gdf_column* col = columns[idx];
+        const char* delim = ((idx+1)<count ? delimiter : terminator);
+        NVStrings* sdata = column_to_strings_csv(col,delim,true_value,false_value);
```

Comment:
So add _size() companion method to go with each of the NVStrings conversion methods?

Reply:
Just a thought. You have to convert the whole column to an nvstrings in order to get its size, and then you allocate the output and convert it again. If that becomes a bottleneck, this might be a possible optimization.